merge.go

```diff
@@ -120,7 +120,12 @@ func (r *mergedRowGroupRows) ReadRows(rows []Row) (n int, err error) {
 				r.err = err
 				return n, err
 			}
-			heap.Pop(r)
+			if c, ok := heap.Pop(r).(rowGroupCursor); ok {
```

Comment:
Maybe we could avoid the type assertion by taking this approach?
```go
evict := r.cursors[0]
heap.Pop(r)
if err := evict.close(); err != nil {
    ...
}
```

Reply:
I want to push back on that. I think the problem with that is that it introduces subtle bugs since it's not explicitly clear that we're closing the cursor that was popped from the heap. Your suggested approach implies the caller knows which cursor was popped from the heap (i.e. how the internal `Pop` implementation works). In fact, the snippet is wrong since the cursor that is actually popped from the heap is the last one in the list (i.e. `s/r.cursors[0]/r.cursors[len(r.cursors-1)]`). I think the type assertion is worth it for more explicit code but also happy to change if you disagree.

Reply:
The heap interface actually documents it:
https://pkg.go.dev/container/heap
> The minimum element in the tree is the root, at index 0.

https://pkg.go.dev/container/heap#Pop
> Pop removes and returns the minimum element (according to Less) from the heap.

There is some ambiguity, since the heap index are not required to map to the slice indexes, but they do in our implementation, so we actually have strong guarantees that the code will remain correct across future Go versions.

I tend to prefer optimizing for compile time type checks, which also has the benefit of removing the runtime cost of interface assertion. I agree it's a trade off, if you want to keep it as-is I'm OK with it.